### PR TITLE
feat(parser)!: Introduction of clap based parser

### PR DESCRIPTION
**Pull request by MartinFillon**
This pr intoduce clap based parser.
Please note that this is still a wip.

Breaking Changes:
-v no longer lists version.
Version is no longer as exhaustive as before.

### CODE REVIEW

**Comment by MartinFillon** [comment]
@cafkafk I just saw that you didn't generated the output of the ptests is this intended ?

**Comment by cafkafk** [comment]
> @cafkafk I just saw that you didn't generated the output of the ptests is this intended ?

uhh, wym?

**Comment by cafkafk** [comment]
> > uhh, wym?
> 
> I cannot find the `.stdout` and `.stderr` from ptests

Should be `tests/gen/ptest_<hash>`

**Comment by MartinFillon** [comment]
Okay my bad I didn't thought of those

**Comment by cafkafk** [comment]
> Okay my bad I didn't thought of those

nw, btw I *think* powertests now cover all options, at least up until v0.15.1, so we are getting very very close

**Review by cafkafk — CHANGES_REQUESTED** [review]
Powertests found an issue with `cargo run -- tests/test_dir --icons always`
```
cargo run -q -- tests/test_dir --icons always
"always": No such file or directory (os error 2)
 git   grid   group   icons   perms   root   size   specials   symlinks   time
```

**Comment by MartinFillon** [comment]
gonna fix this tonight or tomorrow, and finish the unit tests implementations

**Comment by MartinFillon** [comment]
This is ready

File: src/options/parser.rs
Comment (tertsdiepraam) [inline src/options/parser.rs:45]: You can use an alias to use multiple spellings:
```
#[arg(long, alias = "colour")]
color: ...,
```

File: src/options/parser.rs
Comment (tertsdiepraam) [inline src/options/parser.rs:184]: It seems strange to derive `Default` and implement a `default` method outside of `Default`.

**Review by tertsdiepraam — COMMENTED** [review]
Hi! I came across your discussion on matrix and figured I could give some feedback since we have had to deal with complex clap CLIs a lot over at uutils. Of course, I'm not a maintainer here, so feel free to ignore all of this completely. I'm not in a position to decide what implementation is best for eza.

File: src/options/parser.rs
Comment (PThorpe92) [inline src/options/parser.rs:45]: So it seems we might be able to use Enums here instead of a bunch of strings 

https://docs.rs/clap/latest/clap/_derive/_tutorial/chapter_3/index.html#enumerated-values

File: src/main.rs
Comment (PThorpe92) [reply src/main.rs:68]: Didn't realize that other PR for piping args hadn't been merged yet. At the very least it's just something we need to remember to do if it's merged prior to this :+1:

Comment (MartinFillon) [reply src/main.rs:68]: Well it won't compile kinda so I will see it

File: src/options/parser.rs
Comment (PThorpe92) [reply src/options/parser.rs:45]: Didn't realize Terts had already suggested that and provided an example :+1:

File: src/options/parser.rs
Comment (MartinFillon) [reply src/options/parser.rs:43]: I have a problem with this it forces the user to specify a value after the argument and it seems that there is no default_missing_value_t

File: src/options/parser.rs
Comment (MartinFillon) [reply src/options/parser.rs:184]: Well caught

File: src/main.rs
Comment (MartinFillon) [reply src/main.rs:139]: Merge is done

File: src/options/parser.rs
Comment (PThorpe92) [reply src/options/parser.rs:43]: Why couldn't it just default to Auto?

Comment (MartinFillon) [reply src/options/parser.rs:43]: because default_value is for the case the arg was not passed

File: src/options/parser.rs
Comment (tertsdiepraam) [inline src/options/parser.rs:11]: I think I made a comment about this before but maybe I forgot to click the "submit" button because I can't find it anymore. I assume the `ArgAction::Count` is to allow multiple occurrences right? You could get the same behaviour with `#[command(args_override_self = true)]`

See: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b76648135d08a519d66e46ae51d0d63b

Another thing to think about is which argument "wins" if they conflict. As far as I can tell, you don't really have any logic for that. Consider for example `eza -GT` and `eza -TG`.

Comment (tertsdiepraam) [reply src/options/parser.rs:11]: I don't think `clap` allows `arg=true` by default for boolean fields, though? The default `ArgAction` for `bool` is `SetTrue`, which doesn't take an argument. Do you have an example of an argument being verified by count? I'm not entirely sure what you mean by that.

Comment (MartinFillon) [reply src/options/parser.rs:11]: all is verified by count as `-a` and `-a -a` is different, so I thought lets put everything under count in order to be ready if anything like that is tb addeed

Comment (tertsdiepraam) [reply src/options/parser.rs:11]: Right, that makes sense. I still disagree, because I think that the distinction makes the intent of the argument more clear. For example, when you look at a field, you can't see for which arguments the count matters and for which it doesn't. But this is exactly where I defer to the maintainers here for a decision. They might agree with you :)

Comment (MartinFillon) [reply src/options/parser.rs:11]: Well I implemented to match how the original parser was made but still making it kinda easy to understand

**Comment by MartinFillon** [comment]
poke @cafkafk how are we doing on this

**Review by cafkafk — CHANGES_REQUESTED** [review]
> poke @cafkafk how are we doing on this

I suppose first, we'll have to resolve the conflicts with this and main.

**Comment by MartinFillon** [comment]
yh I guess so xd

**Comment by MartinFillon** [comment]
this time its good normally

**Comment by MartinFillon** [comment]
poke @cafkafk @PThorpe92

**Comment by PThorpe92** [comment]
is this 100% ready for final review? @MartinFillon

**Comment by MartinFillon** [comment]
Yes if nothing is merged involving the args

**Comment by cafkafk** [comment]
> Yes if nothing is merged involving the args

I see, I'll hold of for a while merging PRs with arg changes, let's all do that @eza-community/eza-commiters for a little while.

**Review by cafkafk — CHANGES_REQUESTED** [review]
IIRC we might be missing some flags in `powertest.yaml` that have been added since that file was created, we should make sure to also have those. 

I'll slowly try to get my way through the code, but so far it does seem to be working. I also wonder if #564 works with this :thinking:

**Comment by MartinFillon** [comment]
#564 works just see the unit tests that i made to prove it, as per the powertests I am going to run them and check

**Comment by MartinFillon** [comment]
see #741

**Pull request closed without merging**
